### PR TITLE
Improved search algorithm (Course)

### PR DESCRIPTION
**Pull request by bradknaysi**
# What I did
Built a significantly better search algorithm for the `uva-courses` engine using the various query parameters that Elastic App Search API offers. I also added support for engine-specific search algorithms (e.g. course vs instructor).

This adds to the progress of #75

### CODE REVIEW

File: tcf_website/models/models.py [hunk 1]
```diff
@@ -300,6 +300,10 @@ def average_difficulty(self):
         return Review.objects.filter(course=self).aggregate(
             models.Avg('difficulty'))['difficulty__avg']
 
+    def review_count(self):
```
Comment (brian-yu) [inline tcf_website/models/models.py:303]: I believe you could also do `self.review_set.count()`

File: tcf_website/views/search.py [hunk 2]
```diff
@@ -110,3 +108,53 @@ def format_instructors(results):
         formatted.append(instructor)
 
     return formatted
+
+
+def build_elasticsearch_query(query, api_endpoint):
```
Comment (brian-yu) [inline tcf_website/views/search.py:113]: I think we should abstract away the api_endpoints. 

Maybe add `search_courses` and `search_instructors` that wrap `build_elasticsearch_query` and use os.environ directly so we don't have to keep passing around `api_endpoint`? Then lines 21 and 22 would be `search_courses(query)` and `search_instructors(query)`

File: tcf_website/models/models.py (hunk 1 above)
Comment (bradknaysi) [reply tcf_website/models/models.py:303]: I wasn't gonna change it so the code stays consistent w/ the other course model methods

Comment (brian-yu) [reply tcf_website/models/models.py:303]: Hmm I'd argue the other methods (that I wrote haha) should be refactored to use `self.review_set.X` as it seems more readable

File: tcf_website/views/search.py (hunk 2 above)
Comment (bradknaysi) [reply tcf_website/views/search.py]: done

File: tcf_website/models/models.py (hunk 1 above)
Comment (bradknaysi) [reply tcf_website/models/models.py:303]: done

**Pull request merged**